src/test/java/org/elasticsearch/action/support/replication/ShardReplicationOperationTests.java

```diff
@@ -582,5 +811,266 @@ static DiscoveryNode newNode(int nodeId) {
+
+        ActionWithDelay(Settings settings, String actionName, TransportService transportService, ClusterService clusterService, ThreadPool threadPool) throws IOException {
+            super(settings, actionName, transportService, clusterService, threadPool);
+        }
+
+        @Override
+        protected Tuple<Response, Request> shardOperationOnPrimary(ClusterState clusterState, PrimaryOperationRequest shardRequest) throws Throwable {
+            awaitLatch();
+            return new Tuple<>(new Response(), shardRequest.request);
+        }
+
+        private void awaitLatch() throws InterruptedException {
+            countDownLatch.await();
+            countDownLatch = new CountDownLatch(1);
+        }
+
+        @Override
+        protected void shardOperationOnReplica(ReplicaOperationRequest shardRequest) {
+            try {
+                awaitLatch();
+            } catch (InterruptedException e) {
+            }
+        }
+    }
+
+    /*
+    * Transport channel that is needed for replica operation testing.
+    * */
+    public TransportChannel createTransportChannel() {
+        return new TransportChannel() {
+
+            @Override
+            public String action() {
+                return null;
+            }
+
+            @Override
+            public void sendResponse(TransportResponse response) throws IOException {
+            }
+
+            @Override
+            public void sendResponse(TransportResponse response, TransportResponseOptions options) throws IOException {
+            }
+
+            @Override
+            public void sendResponse(Throwable error) throws IOException {
+            }
+        };
+    }
+
+    /**
+     * Creates an IndexShard that can do nothing except for counting operations.
+     */
+    static IndexShard createIndexShard(ShardId shardId, ClusterService clusterService) throws IOException {
+        Index index = shardId.index();
+        // pass the minimum of needed objects to pass initialization
+        return new IndexShard(shardId,
+                new IndexSettingsService(index, ImmutableSettings.EMPTY), null,
+                createStore(shardId),
+                null, new FsTranslog(shardId, ImmutableSettings.EMPTY, createTempDir()), null, null, null, null,
+                null, null,
+                new ShardGetService(shardId, ImmutableSettings.EMPTY, null, null, null),
+                null, null, null, null, null, null, null,
+                new ShardTermVectorsService(shardId, ImmutableSettings.EMPTY, null, null),
+                null, null, null, null, null, null,
+                new SnapshotDeletionPolicy(new KeepOnlyLastDeletionPolicy(shardId, EMPTY_SETTINGS)),
+                new SimilarityService(index), null,
+                null, clusterService, null, null);
+    }
+
+    // from ShadowEngineTests
+    static protected Store createStore(ShardId shardId) throws IOException {
+        return createStore(new RAMDirectory(), shardId);
```

Comment:
if I do that then this is expected to be closed in the end but because this is only used for initializing the fake IndexShard it will never be closed